python/rascaline/rascaline/utils/splines.py

```diff
@@ -0,0 +1,396 @@
+            with np.errstate(divide="ignore"):  # Ignore divide-by-zero warnings
+                mean_relative_error = np.mean(
+                    np.abs((estimated_values - new_values) / new_values)
+                )
+
+            if (
+                mean_absolute_error < self.requested_accuracy
+                or mean_relative_error < self.requested_accuracy
+            ):
+                break
+
+            new_derivatives = self.derivatives_fn(intermediate_positions)
+
+            concatenated_positions = np.concatenate(
+                [self.spline_positions, intermediate_positions], axis=0
+            )
+            concatenated_values = np.concatenate(
+                [self.spline_values, new_values], axis=0
+            )
+            concatenated_derivatives = np.concatenate(
+                [self.spline_derivatives, new_derivatives], axis=0
+            )
+
+            sort_indices = np.argsort(concatenated_positions, axis=0)
+
+            self.spline_positions = concatenated_positions[sort_indices]
+            self.spline_values = concatenated_values[sort_indices]
+            self.spline_derivatives = concatenated_derivatives[sort_indices]
+
+        return self.spline_positions, self.spline_values, self.spline_derivatives
+
+    def _compute_from_spline(self, positions):
+        x = positions
+        delta_x = self.spline_positions[1] - self.spline_positions[0]
+        n = (np.floor(x / delta_x)).astype(np.int32)
+
+        t = (x - n * delta_x) / delta_x
+        t_2 = t**2
+        t_3 = t**3
+
+        h00 = 2.0 * t_3 - 3.0 * t_2 + 1.0
+        h10 = t_3 - 2.0 * t_2 + t
+        h01 = -2.0 * t_3 + 3.0 * t_2
+        h11 = t_3 - t_2
+
+        p_k = self.spline_values[n]
+        p_k_1 = self.spline_values[n + 1]
+
+        m_k = self.spline_derivatives[n]
+        m_k_1 = self.spline_derivatives[n + 1]
+
+        new_shape = (-1,) + (1,) * self.number_of_custom_axes
+        h00 = h00.reshape(new_shape)
+        h10 = h10.reshape(new_shape)
+        h01 = h01.reshape(new_shape)
+        h11 = h11.reshape(new_shape)
+
+        interpolated_values = (
+            h00 * p_k + h10 * delta_x * m_k + h01 * p_k_1 + h11 * delta_x * m_k_1
+        )
+
+        return interpolated_values
+
+
+class SplinerFromFunction(SplinerBase):
+    """Compute the radial integral spline points based on a provided function.
+
+    :parameter radial_integral: Function to compute the radial integral. Function must
+        take ``n``, ``l``, and ``positions`` as inputs, where ``n`` and ``l`` are
+        integers and ``positions`` is a numpy 1-D array that contains the spline points
+        at which the radial integral will be evaluated. The function must return a numpy
+        1-D array containing the values of the radial integral. :parameter cutoff:
+        Spherical cutoff for the radial basis :parameter max_angular: number of radial
+        components
```

Comment:
```suggestion
        1-D array containing the values of the radial integral. 
        :parameter cutoff: Spherical cutoff for the radial basis 
```

Reply:
thanks. Always these issues with the auto formatting.